nntrainer/models/dynamic_training_optimization.cpp

```diff
@@ -0,0 +1,144 @@
+  threshold(threshold_),
+  enabled(false),
+  epsilon(1e-7),
+  skip_n_iterations(skip_n_iter) {
+  reduce_op = reduceByNorm;
+  calc_ratio_op = ratioUsingDerivative;
+  rng.seed(getSeed());
+  dist = std::uniform_real_distribution<float>(0.0, 1.0);
+}
+
+/**
+ * @brief     Check if the given weights can skip updating
+ * @note true if should be applied, else false
+ */
+bool DynamicTrainingOptimization::checkIfApply(
+  const std::vector<Weight> &weights, const std::shared_ptr<Var_Grad> input,
+  const std::shared_ptr<Var_Grad> output, const std::shared_ptr<Optimizer> opt,
+  int iteration) {
+  if (!enabled || iteration < skip_n_iterations)
+    return true;
+
+  std::vector<bool> apply;
+  apply.reserve(weights.size());
+
+  for (auto const &weight : weights)
+    apply.push_back(checkIfApply(weight, input, output, opt, iteration));
+
+  return std::accumulate(apply.begin(), apply.end(), true,
+                         std::logical_and<bool>());
+}
+
+/**
+ * @brief     Check if the given weight can skip updating
+ * @note true if should be applied, else false
+ */
+bool DynamicTrainingOptimization::checkIfApply(
+  const Weight &weight, const std::shared_ptr<Var_Grad> &input,
+  const std::shared_ptr<Var_Grad> &output,
+  const std::shared_ptr<Optimizer> &opt, int iteration) {
+  if (iteration < skip_n_iterations)
+    return true;
+
+  if (!weight.getTrainable() || weight.getGradientRef().uninitialized())
+    return true;
+
+  float reduced_ratio = calc_ratio_op(weight, input, output, reduce_op);
+
+  return checkIfApply(reduced_ratio, (float)opt->getLearningRate(iteration));
+}
+
+/**
+ * @brief   Calculate the ratio of update to the weight using derivative
+ */
+float DynamicTrainingOptimization::ratioUsingDerivative(
+  const Weight &weight, const std::shared_ptr<Var_Grad> &input,
+  const std::shared_ptr<Var_Grad> &output,
+  std::function<float(Tensor const &)> reduce_op) {
+  float reduced_derivative = reduce_op(output->getGradientRef());
+  float reduced_input = reduce_op(input->getVariableRef());
+  float reduced_weight = reduce_op(weight.getVariableRef());
+  float reduced_grad = reduced_derivative * reduced_input;
+
+  return reduced_grad / reduced_weight;
+}
+
+/**
+ * @brief   Calculate the ratio of update to the weight using gradient
+ */
+float DynamicTrainingOptimization::ratioUsingGradient(
+  const Weight &weight, const std::shared_ptr<Var_Grad> &input,
+  const std::shared_ptr<Var_Grad> &output,
+  std::function<float(Tensor const &)> reduce_op) {
+  Tensor ratio = weight.getGradientRef().divide(weight.getVariableRef());
+  return reduce_op(ratio);
+}
+
+/**
+ * @brief   Check if the update should be applied or skipped
+ * @note true if should be applied, else false
+ */
+bool DynamicTrainingOptimization::checkIfApply(float reduced_ratio,
+                                               float learning_rate) {
+  /**
+   * If the reduced update ratio is higher than 1, then always apply update.
```

Comment:
According to the code below it seems that reduced update ratio is higher than 1 would it skip the update

Reply:
This was inverted. Updated the return values.